lib/src/characters_impl.dart

```diff
@@ -400,199 +299,622 @@ class _Characters extends Iterable<String> implements Characters {
-          return _Characters(string.substring(start));
+          return StringCharacters(string.substring(start));
         }
         start = index;
       }
     }
     return this;
   }
 
-  int indexAfter(Characters other, [int startIndex]) {
-    int length = string.length;
-    String otherString = other.string;
-    int otherLength = otherString.length;
-    if (startIndex == null) {
-      startIndex = 0;
-    } else {
-      RangeError.checkValueInInterval(startIndex, 0, length, "startIndex");
+  @override
+  Characters toLowerCase() => StringCharacters(string.toLowerCase());
+
+  @override
+  Characters toUpperCase() => StringCharacters(string.toUpperCase());
+
+  @override
+  bool operator ==(Object other) =>
+      other is Characters && string == other.string;
+
+  @override
+  int get hashCode => string.hashCode;
+
+  @override
+  String toString() => string;
+
+  @override
+  CharacterRange findFirst(Characters characters) {
+    var range = _rangeAll;
+    if (range.collapseToFirst(characters)) return range;
+    return null;
+  }
+
+  @override
+  CharacterRange findLast(Characters characters) {
+    var range = _rangeAll;
+    if (range.collapseToLast(characters)) return range;
+    return null;
+  }
+}
+
+/// A [CharacterRange] on a single string.
+class StringCharacterRange implements CharacterRange {
+  /// The source string.
+  final String _string;
+
+  /// Start index of range in string.
+  ///
+  /// The index is a code unit index in the [String].
+  /// It is always at a grapheme cluster boundary.
+  int _start;
+
+  /// End index of range in string.
+  ///
+  /// The index is a code unit index in the [String].
+  /// It is always at a grapheme cluster boundary.
+  int _end;
+
+  /// The [current] value is created lazily and cached to avoid repeated
+  /// or unnecessary string allocation.
+  String _currentCache;
+
+  StringCharacterRange(String string) : this._(string, 0, 0);
+  StringCharacterRange._(this._string, this._start, this._end);
+
+  /// Changes the current range.
+  ///
+  /// Resets all cached state.
+  void _move(int start, int end) {
+    _start = start;
+    _end = end;
+    _currentCache = null;
+  }
+
+  /// Creates a [Breaks] from [_end] to `_string.length`.
+  ///
+  /// Uses information stored in [_state] for cases where the next
+  /// character has already been seen.
+  Breaks _breaksFromEnd() {
+    return Breaks(_string, _end, _string.length, stateSoTNoBreak);
+  }
+
+  /// Creates a [Breaks] from string start to [_start].
+  ///
+  /// Uses information stored in [_state] for cases where the previous
+  /// character has already been seen.
+  BackBreaks _backBreaksFromStart() {
+    return BackBreaks(_string, _start, 0, stateEoTNoBreak);
+  }
+
+  /// Finds [pattern] in the range from [start] to [end].
+  ///
+  /// Both [start] and [end] are grapheme cluster boundaries in the
+  /// [_string] string.
+  int _indexOf(String pattern, int start, int end) {
```

Comment:
Since start and end are with reference to _string, might as well make this a static or top-level method and pass in _string too. Then you can get rid of the allocations purely to call _indexOf.

Reply:
Good idea.